localstack/services/stepfunctions/asl/component/state/state_execution/state_map/iteration/iterator/distributed_iterator_worker.py

```diff
@@ -0,0 +1,137 @@
+from typing import Optional
+
+from localstack.services.stepfunctions.asl.component.common.error_name.failure_event import (
+    FailureEventException,
+)
+from localstack.services.stepfunctions.asl.component.common.parameters import Parameters
+from localstack.services.stepfunctions.asl.component.common.timeouts.timeout import EvalTimeoutError
+from localstack.services.stepfunctions.asl.component.state.state_execution.state_map.item_selector import (
+    ItemSelector,
+)
+from localstack.services.stepfunctions.asl.component.state.state_execution.state_map.iteration.itemprocessor.map_run_record import (
+    MapRunRecord,
+)
+from localstack.services.stepfunctions.asl.component.state.state_execution.state_map.iteration.iterator.inline_iterator_worker import (
+    InlineIteratorWorker,
+)
+from localstack.services.stepfunctions.asl.component.state.state_execution.state_map.iteration.job import (
+    Job,
+    JobPool,
+)
+from localstack.services.stepfunctions.asl.eval.contextobject.contex_object import Item, Map
+from localstack.services.stepfunctions.asl.eval.environment import Environment
+from localstack.services.stepfunctions.asl.eval.program_state import (
+    ProgramError,
+    ProgramState,
+    ProgramStopped,
+)
+
+
+class DistributedIteratorWorker(InlineIteratorWorker):
+    _map_run_record: MapRunRecord
+
+    def __init__(
+        self,
+        work_name: str,
+        job_pool: JobPool,
+        env: Environment,
+        parameters: Optional[Parameters],
+        map_run_record: MapRunRecord,
+        item_selector: Optional[ItemSelector],
+    ):
+        super().__init__(
+            work_name=work_name,
+            job_pool=job_pool,
+            env=env,
+            parameters=parameters,
+            item_selector=item_selector,
+        )
+        self._map_run_record = map_run_record
+
+    def _eval_job(self, env: Environment, job: Job) -> None:
+        self._map_run_record.item_counter.total.count()
+        self._map_run_record.item_counter.running.count()
+
+        self._map_run_record.execution_counter.total.count()
+        self._map_run_record.execution_counter.running.count()
+
+        job_output = None
+        try:
+            env.context_object_manager.context_object["Map"] = Map(
+                Item=Item(Index=job.job_index, Value=job.job_input)
+            )
+
+            env.inp = job.job_input
+            env.stack.append(env.inp)
+            self._eval_input(env_frame=env)
+
+            job.job_program.eval(env)
+
+            # TODO: verify behaviour with all of these scenarios.
```

Comment:
what would be the strategy to test this? Can this be achieved with certain state machine configuration or would we need to somehow inject errors?

Reply:
Adding some state machines that raise error in the MapRun would inform us about this exact behaviour, but this is very connected to the tolerance control aspects of MapRun which will need supporting from both the provider side and the MapRun backend.